validator/sawtooth_validator/gossip/structure_verifier.py

```diff
@@ -0,0 +1,174 @@
+# Copyright 2017 Intel Corporation
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ------------------------------------------------------------------------------
+
+import logging
+# pylint: disable=import-error,no-name-in-module
+# needed for google.protobuf import
+from google.protobuf.message import DecodeError
+
+from sawtooth_validator.protobuf import client_pb2
+from sawtooth_validator.protobuf.batch_pb2 import BatchHeader
+from sawtooth_validator.protobuf.batch_pb2 import Batch
+from sawtooth_validator.protobuf.block_pb2 import Block
+from sawtooth_validator.protobuf import network_pb2
+from sawtooth_validator.protobuf.network_pb2 import GossipBlockResponse
+from sawtooth_validator.protobuf.network_pb2 import GossipBatchResponse
+from sawtooth_validator.networking.dispatch import Handler
+from sawtooth_validator.networking.dispatch import HandlerResult
+from sawtooth_validator.networking.dispatch import HandlerStatus
+from sawtooth_validator.protobuf.validator_pb2 import Message
+
+
+LOGGER = logging.getLogger(__name__)
+
+
+def validate_block_batches(block):
+    return validate_batches(block.batches)
+
+
+def validate_batches(batches):
+    for batch in batches:
+        if not validate_batch(batch):
+            return False
+
+    return True
+
+
+def validate_batch(batch):
+    # batch structure verification
+    header = BatchHeader()
+    header.ParseFromString(batch.header)
+
+    # validate the transaction_ids field in batch header contains a list of
+    # transaction header_signatures and must be the same order as the
+    # transactions field
+    total = len(batch.transactions)
+
+    if total > len(header.transaction_ids):
+        LOGGER.debug("Batch has extra transactions.Dropping batch: %s",
+                     batch.header_signature)
+        return False
+    elif total < len(header.transaction_ids):
+        LOGGER.debug("Batch lacks transactions.Dropping batch: %s",
+                     batch.header_signature)
+        return False
+
+    index = 0
+    while index < total:
+        txn = batch.transactions[index]
+
+        if header.transaction_ids[index] != txn.header_signature:
+            LOGGER.debug("transaction_ids field does not contains a list of "
+                         "transaction header_signatures or not the same "
+                         "order as the transactions for batch: %s txn: %s",
+                         batch.header_signature,
```

Comment:
This should be reworded, Something like:
The header.transaction_ids does not match the order of transactions in the batch.